src/detect-engine-alert.c

```diff
@@ -337,72 +336,141 @@ static inline void FlowApplySignatureActions(
-            p->alerts.alerts[p->alerts.cnt] = det_ctx->alert_queue[i];
+    /* If there are alerts for this packet, process them */
+    if (max_pos > 0) {
+        PacketAlert tmp_queue[max_pos];
+        uint16_t tmp_size = 0;
+        bool is_action_pass = false;
+        uint32_t pass_sid_num = 0;
+
+        for (uint16_t i = 0; i < max_pos; i++) {
+            const Signature *s = de_ctx->sig_array[det_ctx->alert_queue[i].num];
+            SigThresholdResults res =
+                    PacketAlertFinalize(de_ctx, det_ctx, p, &det_ctx->alert_queue[i], s);
+
+            /* Thresholding removes this alert */
+            if (res == THRESHOLD_SUPPRESSED) {
+                /* we will not copy this to the AlertQueue */
+                p->alerts.suppressed++;
+                SCLogDebug("Suppressing sid %" PRIu32 " alert from alerts' queue", s->id);
+                continue;
+            }
+            tmp_queue[tmp_size] = det_ctx->alert_queue[i];
+            tmp_size++;
             SCLogDebug("Appending sid %" PRIu32 " alert to Packet::alerts at pos %u", s->id, i);
-
-            if (PacketTestAction(p, ACTION_PASS)) {
+            if (PacketTestAction(p, ACTION_PASS) && !is_action_pass) {
                 /* Ok, reset the alert cnt to end in the previous of pass
                  * so we ignore the rest with less prio */
-                break;
+                pass_sid_num = s->num;
+                is_action_pass = true;
             }
-            p->alerts.cnt++;
+        }
+
+        if (packet_alert_max > tmp_size) {
+            p->alerts.cnt = tmp_size;
         } else {
-            p->alerts.discarded++;
+            p->alerts.cnt = packet_alert_max;
+            // we also have discarded alerts in case of queue expansion failure, let's
+            // not miss that count
+            p->alerts.discarded += (tmp_size - packet_alert_max);
+        }
+        qsort(tmp_queue, tmp_size, sizeof(PacketAlert), AlertQueueSortHelper);
+
+        if (det_ctx->is_alert_queue_expand_failure) {
+            if (p->alerts.drop.action & ACTION_DROP) {
+                PacketAlertFinalize(de_ctx, det_ctx, p, &p->alerts.drop, p->alerts.drop.s);
+            }
+        }
+
+        if (is_action_pass) {
+            // this means we got a 'PASS' action for this Packet, pass_sid_num
+            // has the related signature's internal id
+            uint16_t j = 0;
+            for (; j < p->alerts.cnt; j++) {
+                if (tmp_queue[j].num < pass_sid_num) {
+                    // we add it to the final queue
+                    p->alerts.alerts[j] = tmp_queue[j];
+                } else {
+                    break;
+                }
+            }
+            if (PacketTestAction(p, ACTION_DROP)) {
+                if (p->alerts.drop.s->num > pass_sid_num) {
+                    // if this should be a pass, then ignore drop action
+                    p->alerts.drop.action = 0;
+                    PacketPass(p);
```

Comment:
This isn't enough. From my investigations, that's because both the packet and the flow, at this point, have already had the drop action set to them.
I have tried to clear the action with a bitwise operation, but it's not enough, and triggers a bug in libhtp, so I am a bit blocked now.
But overall I am not sure if this is a good approach.

Reply:
Where has the drop already been set?

Reply:
in `AlertQueueAppend()` -> https://github.com/OISF/suricata/blob/master/src/detect-engine-alert.c#L278
doing that there to prevent losing the drop in case there's a failure when trying to append, or to prevent it not being set due to max alert-que size.